Issues with MSVC

I encountered some problems when compiling the last development version with MS Visual Studio 2017 (cl.exe version: 19.16.27024.1):

- Release mode fails with fatal error C1001 at `source/fe/mapping.cc`, line 58. Enclosing function `project_real_point_to_unit_point_on_face` with `#pragma optimize("", off)` and `#pragma optimize("d", on)` solves the problem. I cann't figure out why but it just works.
- Configure fails at `cmake/config/CMakeLists.txt`, line 286 when configed with intel mkl libraries. Since pkgconfig files are useless on Windows, I just delete these lines.
- Linking error when compiling Step-16 (maybe other multigrid related tutorials) due to  unresolved external symbol `dealii::internal::MatrixFreeFunctions::ShapeInfo::reinit`, which is called in `dealii::internal::MGTransfer::setup_element_info`. Why use [this](https://github.com/dealii/dealii/blob/c89e1e740ad8b32def0e490cd4b5e62dd164a4b1/source/matrix_free/matrix_free.inst.in#L85)?
- Some minor warnings, e.g. wd4804, wd4819.

The output are as follows:

```
-- Configuring done. Proceed to target definitions now.
-- Setting up bundled features
-- BOOST::Iostreams will not be available because zlib or its header files could not be found
-- Performing Test DEAL_II_HAVE_FLAG_Wno_sign_compare
-- Performing Test DEAL_II_HAVE_FLAG_Wno_sign_compare - Failed
-- Performing Test DEAL_II_HAVE_FLAG_Wno_write_strings
-- Performing Test DEAL_II_HAVE_FLAG_Wno_write_strings - Failed
-- Performing Test DEAL_II_HAVE_FLAG_Wno_switch
-- Performing Test DEAL_II_HAVE_FLAG_Wno_switch - Failed
-- Performing Test DEAL_II_HAVE_FLAG_Wno_nested_anon_types
-- Performing Test DEAL_II_HAVE_FLAG_Wno_nested_anon_types - Failed
-- Performing Test DEAL_II_HAVE_FLAG_Wno_implicit_fallthrough
-- Performing Test DEAL_II_HAVE_FLAG_Wno_implicit_fallthrough - Failed
-- Performing Test DEAL_II_HAVE_FLAG_Wno_cast_function_type
-- Performing Test DEAL_II_HAVE_FLAG_Wno_cast_function_type - Failed
-- Setting up bundled features - Done
-- Setting up library
-- Setting up library - Done
-- Setting up project configuration
CMake Error at cmake/config/CMakeLists.txt:286 (STRING):
  STRING sub-command REGEX, mode REPLACE: regex "^(|)" matched an empty
  string.


CMake Error at cmake/config/CMakeLists.txt:286 (STRING):
  STRING sub-command REGEX, mode REPLACE: regex "^(|)" matched an empty
  string.


CMake Error at cmake/config/CMakeLists.txt:286 (STRING):
  STRING sub-command REGEX, mode REPLACE: regex "^(|)" matched an empty
  string.

-- Setting up project configuration - Done
-- Setting up examples
-- Setting up examples - Done
-- Setting up quick_tests in RELEASE mode
-- Setting up quick_tests in RELEASE mode - Done
-- Setting up testsuite
-- Setting up testsuite - Done
###
#
#  deal.II configuration:
#        CMAKE_BUILD_TYPE:       Release
#        BUILD_SHARED_LIBS:      OFF
#        CMAKE_INSTALL_PREFIX:   D:/libs/dealii-x64
#        CMAKE_SOURCE_DIR:       D:/tmp/sources/dealii
#                                (version 9.1.0-pre, shortrev c89e1e740a)
#        CMAKE_BINARY_DIR:       D:/tmp/sources/dealii/build-vs-x64
#        CMAKE_CXX_COMPILER:     MSVC 19.16.27024.1 on platform Windows AMD64
#                                C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/cl.exe
#
#  Configured Features (DEAL_II_ALLOW_BUNDLED = ON, DEAL_II_ALLOW_AUTODETECTION = ON):
#      ( DEAL_II_WITH_64BIT_INDICES = OFF )
#      ( DEAL_II_WITH_ADOLC = OFF )
#      ( DEAL_II_WITH_ARPACK = OFF )
#      ( DEAL_II_WITH_ASSIMP = OFF )
#        DEAL_II_WITH_BOOST set up with bundled packages
#        DEAL_II_WITH_COMPLEX_VALUES = ON
#      ( DEAL_II_WITH_CUDA = OFF )
#      ( DEAL_II_WITH_CXX14 = OFF )
#      ( DEAL_II_WITH_CXX17 = OFF )
#      ( DEAL_II_WITH_GMSH = OFF )
#      ( DEAL_II_WITH_GSL = OFF )
#      ( DEAL_II_WITH_HDF5 = OFF )
#        DEAL_II_WITH_LAPACK set up with external dependencies
#      ( DEAL_II_WITH_METIS = OFF )
#      ( DEAL_II_WITH_MPI = OFF )
#        DEAL_II_WITH_MUPARSER set up with bundled packages
#        DEAL_II_WITH_NANOFLANN set up with external dependencies
#      ( DEAL_II_WITH_NETCDF = OFF )
#      ( DEAL_II_WITH_OPENCASCADE = OFF )
#      ( DEAL_II_WITH_P4EST = OFF )
#      ( DEAL_II_WITH_PETSC = OFF )
#      ( DEAL_II_WITH_SCALAPACK = OFF )
#      ( DEAL_II_WITH_SLEPC = OFF )
#      ( DEAL_II_WITH_SUNDIALS = OFF )
#      ( DEAL_II_WITH_THREADS = OFF )
#      ( DEAL_II_WITH_TRILINOS = OFF )
#        DEAL_II_WITH_UMFPACK set up with bundled packages
#      ( DEAL_II_WITH_ZLIB = OFF )
#
#  Component configuration:
#      ( DEAL_II_COMPONENT_DOCUMENTATION = OFF )
#        DEAL_II_COMPONENT_EXAMPLES
#      ( DEAL_II_COMPONENT_PACKAGE = OFF )
#      ( DEAL_II_COMPONENT_PYTHON_BINDINGS = OFF )
#
#  Detailed information (compiler flags, feature configuration) can be found in detailed.log
#
#  Run  $ make info  to print a help message with a list of top level targets
#
###
-- Configuring incomplete, errors occurred!
See also "D:/tmp/sources/dealii/build-vs-x64/CMakeFiles/CMakeOutput.log".
See also "D:/tmp/sources/dealii/build-vs-x64/CMakeFiles/CMakeError.log".
```

The command I used to configure are
```
cmake -G "NMake Makefiles" ^
      -DCMAKE_BUILD_TYPE=Release ^
      -DCMAKE_INSTALL_PREFIX=D:/libs/dealii-x64 ^
      -DLAPACK_FOUND=TRUE -DLAPACK_LIBRARIES="D:/libs/intel-mkl-x64/lib/mkl_intel_lp64_dll.lib;D:/libs/intel-mkl-x64/lib/mkl_sequential_dll.lib;D:/libs/intel-mkl-x64/lib/mkl_core_dll.lib" ^
      -DDEAL_II_WITH_NANOFLANN=ON -DNANOFLANN_DIR="D:/libs/nanoflann" ^
      ..
```

```
CMake Error at cmake/config/CMakeLists.txt:286 (STRING): 
STRING sub-command REGEX, mode REPLACE: regex "^(|)" matched an empty string. 
```

this has nothing to do with `pkgconfig` (a part from that the error occurs computing variables used by him), it is `cmake` itself that is complaining. It looks like the variables `CMAKE_STATIC_LIBRARY_PREFIX` and `CMAKE_SHARED_LIBRARY_PREFIX` are not defined. @tamiko, do you have any idea? BTW, I am leaning to disable the pkg-config path if we are on windows

@adamqc, which version of `cmake` do you have?

--- a/cmake/config/CMakeLists.txt
+++ b/cmake/config/CMakeLists.txt
@@ -247,6 +247,28 @@ ENDFOREACH()
 #
 
 #
+# We need to gather some variables for the regex below to work.
+#
+
+SET(_library_prefixes "")
+IF (CMAKE_SHARED_LIBRARY_PREFIX)
+  LIST(APPEND _library_prefixes ${CMAKE_SHARED_LIBRARY_PREFIX})
+ENDIF()
+IF (CMAKE_STATIC_LIBRARY_PREFIX)
+  LIST(APPEND _library_prefixes ${CMAKE_STATIC_LIBRARY_PREFIX})
+ENDIF()
+STRING(REPLACE ";" "|" _library_prefixes "${_library_prefixes}")
+
+SET(_library_suffixes "")
+IF (CMAKE_SHARED_LIBRARY_SUFFIX)
+  LIST(APPEND _library_suffixes ${CMAKE_SHARED_LIBRARY_SUFFIX})
+ENDIF()
+IF (CMAKE_STATIC_LIBRARY_SUFFIX)
+  LIST(APPEND _library_suffixes ${CMAKE_STATIC_LIBRARY_SUFFIX})
+ENDIF()
+STRING(REPLACE ";" "|" _library_suffixes "${_library_suffixes}")
+
+#
 # Build up the link line from our list of libraries:
 #
 
@@ -283,14 +305,12 @@ FOREACH(_build ${DEAL_II_BUILD_TYPES})
       ENDIF()
 
       # Recover short name:
-      STRING(REGEX REPLACE
-        "^(${CMAKE_STATIC_LIBRARY_PREFIX}|${CMAKE_SHARED_LIBRARY_PREFIX})"
-        "" _name "${_name}"
-        )
-      STRING(REGEX REPLACE
-        "(${CMAKE_STATIC_LIBRARY_SUFFIX}|${CMAKE_SHARED_LIBRARY_SUFFIX})$"
-        "" _name "${_name}"
-        )
+      IF(_library_prefixes)
+        STRING(REGEX REPLACE "^(${_library_prefixes})" "" _name "${_name}")
+      ENDIF()
+      IF(_library_suffixes)
+        STRING(REGEX REPLACE "(${_library_suffixes})$" "" _name "${_name}")
+      ENDIF()
       SET(_library_string "${_library_string}-l${_name}")
     ENDIF()
 

